risc0/zkp/rust/src/field/goldilocks.rs

```diff
@@ -0,0 +1,745 @@
+    use rand::SeedableRng;
+
+    #[test]
+    /// Roots of unity tests common to all fields under test
+    pub fn roots_of_unity() {
+        field::test::test_roots_of_unity::<Elem>();
+    }
+
+    #[test]
+    pub fn field_ops() {
+        field::test::test_field_ops::<Elem>(P);
+    }
+
+    #[test]
+    pub fn create_element_no_wrap() {
+        let test_element = Elem::from(P - 1u64);
+        let lhs: u64 = test_element.into();
+        assert_eq!(lhs, P - 1u64);
+    }
+
+    #[test]
+    pub fn create_element_field_wrap() {
+        let test_element = Elem::from(P + 1u64);
+        let lhs: u64 = test_element.into();
+        assert_eq!(lhs, 1u64);
+    }
+
+    #[test]
+    fn isa_field() {
+        // Pick random sets of 3 elements of ExtElem, and verify they meet the
+        // requirements of a field.
+        let mut rng = rand::rngs::SmallRng::seed_from_u64(2);
+        for _ in 0..1_000 {
+            let a = ExtElem::random(&mut rng);
+            let b = ExtElem::random(&mut rng);
+            let c = ExtElem::random(&mut rng);
+            // Addition + multiplication commute
+            assert_eq!(a + b, b + a);
+            assert_eq!(a * b, b * a);
+            // Addition + multiplication are associative
+            assert_eq!(a + (b + c), (a + b) + c);
+            assert_eq!(a * (b * c), (a * b) * c);
+            // Distributive property
+            assert_eq!(a * (b + c), a * b + a * c);
+            // Inverses
+            if a != ExtElem::ZERO {
+                assert_eq!(a.inv() * a, ExtElem::from(1));
+            }
+            assert_eq!(ExtElem::ZERO - a, -a);
+            assert_eq!(a + (-a), ExtElem::ZERO);
+        }
+    }
+
+    #[test]
+    fn inv() {
+        // Smoke test for inv
+        assert_eq!(Elem(5).inv() * Elem(5), Elem(1));
+    }
+
+    #[test]
+    fn pow() {
+        // Smoke tests for pow
+        assert_eq!(Elem(5).pow(0), Elem(1));
+        assert_eq!(Elem(5).pow(1), Elem(5));
+        assert_eq!(Elem(5).pow(2), Elem(25));
+        // Mathematica says PowerMod[5, 1000, 2^64 - 2^32 + 1] == 1298979347292407023
+        assert_eq!(Elem(5).pow(1000), Elem(1298979347292407023));
+        assert_eq!(Elem(5).pow((P - 2) as usize) * Elem(5), Elem(1));
+        assert_eq!(Elem(5).pow((P - 1) as usize), Elem(1));
+    }
+
+    #[test]
+    fn compare_native_add_easy() {
+        // Test addition when 0 < a + b < P
+        let fa = Elem::from(2u64);
+        let fb = Elem::from(2u64);
+        let a: u64 = fa.into();
+        let b: u64 = fb.into();
+        assert_eq!(fa + fb, Elem::from(a.wrapping_add(b)));
+    }
+
+    #[test]
+    fn compare_native_add_subfield_wrap() {
+        // Test for addition when P < a + b < 2^64
+        let fa = Elem::from(P - 2u64);
+        let fb = Elem::from(5u64);
+        let a: u64 = fa.into();
+        let b: u64 = fb.into();
+        assert_eq!(fa + fb, Elem::from(a.wrapping_add(b)));
+    }
+
+    #[test]
+    fn compare_native_add_u64_wrap() {
```

Comment:
Am I missing something or is this the same test as `compare_native_add_subfield_wrap`?